src/R/Components/Impl/Application/Configuration/Parser/ConfigurationParser.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Microsoft.Common.Core;
+using Microsoft.R.Core.AST.Operators;
+using Microsoft.R.Core.AST.Operators.Definitions;
+using Microsoft.R.Core.AST.Scopes;
+using Microsoft.R.Core.AST.Statements.Definitions;
+using Microsoft.R.Core.AST.Variables;
+using Microsoft.R.Core.Parser;
+using static System.FormattableString;
+
+namespace Microsoft.R.Components.Application.Configuration.Parser {
+    internal sealed class ConfigurationParser {
+        private readonly List<ConfigurationError> _errors = new List<ConfigurationError>();
+        private readonly StreamReader _sr;
+        private string _bufferedLine;
+        private int _lineNumber;
+
+        public IReadOnlyList<ConfigurationError> Errors => _errors;
+
+        public ConfigurationParser(StreamReader sr) {
+            _sr = sr;
+        }
+
+        public IConfigurationSetting ReadSetting() {
+            var setting = new ConfigurationSetting();
+            while (!IsEndOfStream()) {
+                var line = ReadLine();
+                if (!string.IsNullOrWhiteSpace(line)) {
+                    if (line.TrimStart()[0] == '#') {
+                        ReadAttributeValue(line.Substring(1), setting);
+                    } else {
+                        // Expression can be multi-line so read all lines up to the next comment
+                        int startingLineNumber = _lineNumber;
+                        var text = line + ReadRemainingExpressionText();
+                        if (ParseSetting(text, startingLineNumber, setting)) {
+                            return setting;
+                        }
+                        break;
+                    }
+                }
+            }
+            return null;
+        }
+
+        private bool IsEndOfStream() {
+            return _bufferedLine == null && _sr.EndOfStream;
+        }
+        private string ReadLine() {
+            var line = _bufferedLine ?? _sr.ReadLine();
+            _lineNumber++;
```

Comment:
since UnreadLine ```--``` line number, we should be increasing it here in any case i think.